pkg/egresspolicy/manager.go

```diff
@@ -0,0 +1,461 @@
+		}
+
+		// Keep id to pod data mapping up to date
+		manager.podDataStore[podId] = data
+		return data, nil
+	}
+
+	data := &podMetadata{
+		ips:      podIPs,
+		labels:   pod.GetLabels(),
+		nsLabels: namespace.GetLabels(),
+		id:       podId,
+	}
+
+	// Keep id to pod data mapping up to date
+	manager.podDataStore[podId] = data
+	return data, nil
+}
+
+func (manager *Manager) getNamespaceForPod(pod *slim_corev1.Pod) (*slim_corev1.Namespace, error) {
+	nsName := pod.GetNamespace()
+	nsStore := manager.storeGetter.GetStore("namespace")
+	nsItem, exists, err := nsStore.GetByKey(pod.GetNamespace())
+	if err != nil {
+		log.WithError(err).Warn("Failed to get namespace")
+		return nil, err
+	}
+	if !exists {
+		log.WithFields(logrus.Fields{"nsName": nsName}).Warn("Namespace does not exist")
+		return nil, nil
+	}
+	return nsItem.(*slim_corev1.Namespace), nil
+}
+
+// isValidConfig validates the given policy config.
+func (manager *Manager) isValidConfig(config EgressPolicyConfig) error {
+	for _, policyConfig := range manager.policyConfigs {
+		if policyConfig.egressIP.String() == config.egressIP.String() {
+			return fmt.Errorf("CiliumEgressNatPolicy for egress ip %v already exists.", config.egressIP.String())
+		}
+	}
+	return nil
+}
+
+// upsertPolicyPod updates or insert to pod policy mapping for given policy config and pods,
+// it also upserts the pod data catch and egress map to keep in sync
+func (manager *Manager) upsertPolicyPod(config *EgressPolicyConfig, pods ...*podMetadata) error {
+	for _, pod := range pods {
+		podIp := net.ParseIP(pod.ips[0])
+		if podPolicies, ok := manager.policyPods[pod.id]; ok {
+			newPodPolicy := true
+			for _, poID := range podPolicies {
+				// Existing pod policy update, noop for this policy
+				if poID == config.id {
+					newPodPolicy = false
+					break
+				}
+			}
+			if newPodPolicy {
+				// Pod selected by a new policy
+				err := manager.updateEgressMap(podIp, config)
+				if err != nil {
+					return err
+				}
+				manager.policyPods[pod.id] = append(manager.policyPods[pod.id], config.id)
+				manager.podDataStore[pod.id] = pod
+			}
+		} else {
+			// Pod selected by a policy for the first time
+			err := manager.updateEgressMap(podIp, config)
+			if err != nil {
+				return err
+			}
+			pp := []policyID{config.id}
+			manager.policyPods[pod.id] = pp
+			manager.podDataStore[pod.id] = pod
+		}
+	}
+	return nil
+}
+
+func (manager *Manager) updateEgressMap(sip net.IP, config *EgressPolicyConfig) error {
+	var err error
+	for _, dstCIDR := range config.dstCIDRs {
+		key := egressmap.NewKey(sip, dstCIDR.IP, dstCIDR.Mask)
+		value := &egressmap.EgressInfo4{}
+		// As currently designed, the egressIp serves two purposes, one for forwarding traffic
+		// to the gateway node, the other for SNATing the egress traffic on the gateway.
+		copy(value.TunnelEndpoint[:], config.egressIP)
+		copy(value.EgressIP[:], config.egressIP)
+
+		err = egressmap.EgressMap.Update(&key, value)
+		if err != nil {
```

Comment:
Following the above comment:

```suggestion
		if err := egressmap.EgressMap.Update(&key, value); err != nil {
```

Reply:
Fixed.